lib/winapi/sync.c

```diff
@@ -38,6 +40,141 @@ VOID LeaveCriticalSection (LPCRITICAL_SECTION lpCriticalSection)
     RtlLeaveCriticalSection(lpCriticalSection);
 }
 
+#define INITONCE_MASK (((uintptr_t)1 << INIT_ONCE_CTX_RESERVED_BITS) - 1)
+#define INITONCE_UNINITIALIZED 0
+#define INITONCE_IN_PROGRESS 1
+#define INITONCE_ASYNC_IN_PROGRESS 2
+#define INITONCE_DONE 3
+
+BOOL InitOnceBeginInitialize (LPINIT_ONCE lpInitOnce, DWORD dwFlags, PBOOL fPending, LPVOID *lpContext)
+{
+    assert(fPending != NULL);
+
+    if (dwFlags & INIT_ONCE_CHECK_ONLY) {
+        if (dwFlags & INIT_ONCE_ASYNC) {
+            *fPending = FALSE;
+            SetLastError(ERROR_INVALID_PARAMETER);
+            return FALSE;
+        }
+        if (((ULONG_PTR)lpInitOnce->Ptr & INITONCE_MASK) != INITONCE_DONE) {
+            *fPending = TRUE;
+            SetLastError(ERROR_GEN_FAILURE);
+            return FALSE;
+        }
+
+        if (lpContext) {
+            *lpContext = (LPVOID)((ULONG_PTR)lpInitOnce & ~INITONCE_MASK);
```

Comment:
Shouldn't this be `lpInitOnce->Ptr`? This feels like a bug that would even break the implementation?

Reply:
Ouch. Yes, that should be `lpInitOnce->Ptr` instead of just `lpInitOnce`. The mask is correct though.